[CASSANDRA-14556] Optimize Streaming

dineshjoshi opened the pull request:
DTests here - https://github.com/apache/cassandra-dtest/compare/master...dineshjoshi:faster-streaming-rev2

dineshjoshi committed:
Optimize Streaming

dineshjoshi committed:
Add missing file

dineshjoshi committed:
Clean up

dineshjoshi committed:
Fix resource leak warning

dineshjoshi committed:
Circle CI

dineshjoshi committed:
Fix unit tests

dineshjoshi committed:
Enable dtests for CircleCI

dineshjoshi committed:
Use a synchronized map instead of concurrent map to preserve order

aweisberg commented on src/java/org/apache/cassandra/db/streaming/CassandraBlockStreamReader.java:
```diff
@@ -0,0 +1,172 @@
+
+        LifecycleTransaction txn = CassandraStreamReceiver.fromReceiver(session.getAggregator(tableId)).getTransaction();
+
+        // TODO: Is this the correct directory?
```
Is this still TODO?

This still doesn't really look right to me. First off we know the size of the sstable, but aren't providing it to Directories, but secondly we know the range of the sstable and presumably it's going to fall into disk boundaries like flushing? How does this interact with partitioning data per disk? Does it violate assumptions made there? @krummas WDYT?

krummas replied:
Yeah this would violate CASSANDRA-6696 since it writes the whole file to the directory with the most space. Things would still work though, compaction will drop the tokens in the correct place. We have the same problem with nodetool refresh/import.

We could have the sending side send the first token of the sstable so that the receiver can drop the received sstable in the correct directory, which would make it correct if the nodes have the same disk count. 

CASSANDRA-10540 will make it work since the sstables would fit in the requested ranges.

dineshjoshi replied:
Based on @krummas feedback, I have implemented the fix.

aweisberg commented on src/java/org/apache/cassandra/io/util/RebufferingInputStream.java:
```diff
@@ -26,7 +26,9 @@
 import java.nio.ByteOrder;
 
 import net.nicoulaj.compilecommand.annotations.DontInline;
+import org.apache.cassandra.io.FSError;
```
Rando new imports?

dineshjoshi replied:
Fixed

aweisberg commented on src/java/org/apache/cassandra/net/async/ByteBufDataOutputStreamPlus.java:
```diff
@@ -135,6 +140,53 @@ public ChannelFuture writeToChannel(ByteBuffer buffer) throws IOException
+            }
+
+            return bytesTransferred;
+        } catch (Exception e)
```
This is a little odd on that this function doesn't open the file, but it does close it. That's not the normal convention. I think it just needs to be documented that it will take ownership of the file.

dineshjoshi replied:
I have it in the comment - ' * Closes file channel when done'.

aweisberg commented on src/java/org/apache/cassandra/net/async/RebufferingByteBufDataInputPlus.java:
```diff
@@ -249,4 +253,50 @@ public ByteBufAllocator getAllocator()
+
+            int toCopy = (int) Math.min(len - copied, remaining);
+
+            ByteBuffer dup = buffer.duplicate();
```
I don't see why you would store position on the stack outside of the buffer it really just seems like an opportunity to get it wrong by not going direct to the buffer. Same with remaining it's only ever used one after it is ever stored on the stack so why store it on the stack at all?
I hate to keep bikeshedding this, but I don't even see why you need the dup or to manually update the position? 

```
if (buffer.remaining() == 0)
{
    try 
   {
        reBuffer();
    }
    catch (EOFException e)
    {
        throw new EOFException("EOF after " + copied + " bytes out of " + len);
    }
    if (buffer.remaining() == 0)
    {
        return copied == 0 ? -1 : copied;
    }
}

int originalLimit = buffer.limit();
buffer.limit(Math.min(len - copied, remaining);
int written = writer.applyToChannel(c -> c.write(buffer));
buffer.limit(originalLimit);
copied += written;
```

aweisberg commented on src/java/org/apache/cassandra/net/async/RebufferingByteBufDataInputPlus.java:
```diff
@@ -249,4 +253,50 @@ public ByteBufAllocator getAllocator()
+                try
+                {
+                    reBuffer();
+                } catch (EOFException e)
```
This doesn't follow Cassandra's code style, the catch needs its own line.

dineshjoshi replied:
Fixed

aweisberg commented on src/java/org/apache/cassandra/net/async/RebufferingByteBufDataInputPlus.java:
```diff
@@ -249,4 +253,50 @@ public ByteBufAllocator getAllocator()
+                }
+                position = buffer.position();
+                remaining = buffer.remaining();
+                if (remaining == 0)
```
This also doesn't follow the code style, we are supposed to use braces for single line if all the time (from the Sun Java code style).

Can you review the entire change set for style issues?

dineshjoshi replied:
The codebase that I've been through doesn't use braces for single statement blocks. I assumed that was the convention. I can change it in my PR to follow the Sun style but we should ideally audit and change the code style in the future to avoid confusion.

belliottsmith replied:
I think we need to raise this on the mailing list to possibly modify the code style, because for a long time it has been standard practice for many contributors (myself included) to elide the braces for single line statements, because otherwise the code becomes illegible.

Note this all springs from the modification to the sun code style, i.e. the new line per brace.  Everyone would prefer braces, but they would prefer the code to be readable too :/

belliottsmith replied:
Even you've done it!  :)

https://github.com/aweisberg/cassandra/commit/1491a40b7b4ea2723bcf22d870ee514b47ea901b#diff-baae9b53db48ff1ed346e469f7506545R192

jeffjirsa replied:
Style guide says

“Always include braces for nested levels of conditionals and loops. Only avoid braces for single level.”

Skipping for single level has always been ok.

aweisberg replied:
My bad. I missed the last part of it.

And me doing something is not an endorsement of it being a good idea/right!

aweisberg commented on src/java/org/apache/cassandra/service/StorageService.java:
```diff
@@ -4236,7 +4236,7 @@ private void move(Token newToken) throws IOException
 
     private class RangeRelocator
     {
-        private final StreamPlan streamPlan = new StreamPlan(StreamOperation.RELOCATION);
```
Redundant addition of false to the constructor? It's still here.

aweisberg commented on src/java/org/apache/cassandra/io/sstable/format/big/BigTableBlockWriter.java:
```diff
@@ -0,0 +1,222 @@
+                                                                           new HashSet(components).removeAll(supportedComponents));
+
+        for (Component c : components)
+            componentWriters.put(c.type, makeWriter(descriptor, c, writerOption));
```
Missing braces

aweisberg commented on src/java/org/apache/cassandra/io/sstable/format/big/BigTableBlockWriter.java:
```diff
@@ -0,0 +1,222 @@
+    public Collection<SSTableReader> finished()
+    {
+        if (finalReader == null)
+            finalReader = SSTableReader.open(descriptor,
```
Missing braces

aweisberg commented on src/java/org/apache/cassandra/io/sstable/format/big/BigTableBlockWriter.java:
```diff
@@ -0,0 +1,222 @@
+    public Throwable commit(Throwable accumulate)
+    {
+        for (SequentialWriter writer : componentWriters.values())
+            writer.commit(accumulate);
```
missing braces

aweisberg commented on src/java/org/apache/cassandra/io/sstable/format/big/BigTableBlockWriter.java:
```diff
@@ -0,0 +1,222 @@
+    public Throwable abort(Throwable accumulate)
+    {
+        for (SequentialWriter writer : componentWriters.values())
+            writer.abort(accumulate);
```
missing braces

aweisberg commented on src/java/org/apache/cassandra/io/sstable/format/big/BigTableBlockWriter.java:
```diff
@@ -0,0 +1,222 @@
+    @Override
+    public void prepareToCommit()
+    {
+        for (SequentialWriter writer : componentWriters.values())
```
missing braces

aweisberg commented on src/java/org/apache/cassandra/io/sstable/format/big/BigTableBlockWriter.java:
```diff
@@ -0,0 +1,222 @@
+    @Override
+    public void close()
+    {
+        for (SequentialWriter writer : componentWriters.values())
```
missing braces

aweisberg commented on src/java/org/apache/cassandra/io/sstable/format/big/BigTableBlockWriter.java:
```diff
@@ -0,0 +1,222 @@
+    public void writeComponent(Component.Type type, DataInputPlus in, long size)
+    {
+        logger.info("Writing component {} to {} length {}", type, componentWriters.get(type).getPath(), size);
+        if (in instanceof RebufferingByteBufDataInputPlus)
```
missing braces

aweisberg commented on src/java/org/apache/cassandra/io/sstable/format/big/BigTableBlockWriter.java:
```diff
@@ -0,0 +1,222 @@
+        {
+            long bytesWritten = in.consumeUntil(writer, size);
+
+            if (bytesWritten != size)
```
missing braces

aweisberg commented on src/java/org/apache/cassandra/io/util/CheckedFunction.java:
```diff
@@ -0,0 +1,24 @@
+package org.apache.cassandra.io.util;
+
+@FunctionalInterface
+public interface CheckedFunction<T, R, E extends Exception> {
```
newline before brace

aweisberg commented on src/java/org/apache/cassandra/io/sstable/Component.java:
```diff
@@ -74,6 +81,20 @@ static Type fromRepresentation(String repr)
+            return idToType.getOrDefault(id, CUSTOM);
+        }
+
+        static {
```
newline before brace

aweisberg commented on src/java/org/apache/cassandra/db/streaming/CassandraOutgoingFile.java:
```diff
@@ -114,13 +153,59 @@ public void write(StreamSession session, DataOutputStreamPlus out, int version)
+    private boolean fullyContainedIn(List<Range<Token>> requestedRanges, SSTableReader sstable)
+    {
+        if (requestedRanges == null)
+            return false;
```
missing braces

aweisberg commented on src/java/org/apache/cassandra/db/streaming/CassandraOutgoingFile.java:
```diff
@@ -114,13 +153,59 @@ public void write(StreamSession session, DataOutputStreamPlus out, int version)
+                }
+
+                if (foundFlag == false)
+                    return false;
```
missing braces

aweisberg commented on src/java/org/apache/cassandra/db/streaming/CassandraOutgoingFile.java:
```diff
@@ -62,7 +88,20 @@ public CassandraOutgoingFile(StreamOperation operation, Ref<SSTableReader> ref,
+        {
+            File file = new File(sstable.descriptor.filenameFor(component));
+            if (file.exists())
+                result.add(new ComponentInfo(component.type, file.length()));
```
missing braces

aweisberg commented on src/java/org/apache/cassandra/db/streaming/CassandraBlockStreamReader.java:
```diff
@@ -0,0 +1,172 @@
+    {
+        long size = 0;
+        for (ComponentInfo component : components)
+            size += component.length;
```
missing braces

aweisberg commented on src/java/org/apache/cassandra/db/streaming/CassandraBlockStreamReader.java:
```diff
@@ -0,0 +1,172 @@
+    {
+        Directories.DataDirectory localDir = cfs.getDirectories().getWriteableLocation(totalSize);
+        if (localDir == null)
+            throw new IOException(String.format("Insufficient disk space to store %s", FBUtilities.prettyPrintMemory(totalSize)));
```
missing braces

aweisberg commented on test/unit/org/apache/cassandra/io/sstable/ComponentTest.java:
```diff
@@ -0,0 +1,41 @@
+    public void testTypeCreationFromRepresentation_ValidId()
+    {
+        for (int i = 1; i < 10; i++)
+            assertTrue(Component.Type.fromRepresentation((byte) i) != Component.Type.CUSTOM);
```
missing braces

aweisberg commented on test/unit/org/apache/cassandra/io/sstable/format/big/BigTableBlockWriterTest.java:
```diff
@@ -0,0 +1,196 @@
+        for (int i = 0; i < store.metadata().params.minIndexInterval; i++)
+        {
+            DecoratedKey key = Util.dk(String.valueOf(i));
+            if (firstKey == null)
```
Missing braces

aweisberg commented on src/java/org/apache/cassandra/net/async/ByteBufDataOutputStreamPlus.java:
```diff
@@ -135,6 +140,53 @@ public ChannelFuture writeToChannel(ByteBuffer buffer) throws IOException
+                bytesTransferred += toRead;
+                final boolean shouldClose = (bytesTransferred == length); // this is the last buffer, can safely close channel
+
+                channel.writeAndFlush(fileRegion).addListener(future -> {
```
Is the contract here that the entire region will be sent or an error generated? I just want to make sure we don't need to check for the actual amount transferred.

dineshjoshi replied:
Yes, AFAICT, netty guarantees that the whole region will be transferred or an error generated. We don't need to check the number of bytes transferred as we have handed off the file region to the channel.

dineshjoshi committed:
Code review comments

dineshjoshi committed:
Simplify consumeUntil

dineshjoshi committed:
Fix data directory selection

dineshjoshi committed:
update circleci config to point to dtest fork

dineshjoshi committed:
Fix serialized size

dineshjoshi committed:
Testing

dineshjoshi commented:
I'm running into some dtest failures, I am in the process of fixing them.

dineshjoshi committed:
Fix directory selection

dineshjoshi committed:
Readd test

dineshjoshi committed:
Dont use DatabaseDescriptor to determine Partitioner

dineshjoshi committed:
Add tests for consumeUntil

dineshjoshi committed:
Update BTBW test

dineshjoshi committed:
Fix bug in size calculation during benchmark

dineshjoshi committed:
Clean up unused imports

dineshjoshi committed:
Add tests for fullyContainedIn logic

dineshjoshi committed:
Fix comment

dineshjoshi committed:
Update benchmark

dineshjoshi committed:
Clean up

dineshjoshi committed:
Add documentation to cassandra.yaml

dineshjoshi committed:
Update CHANGES.txt

dineshjoshi committed:
Update NEWS.txt

dineshjoshi committed:
Fix typo

dineshjoshi commented:
1. Fixed the data directory selection
2. Added & updated unit tests
3. Added documentation in cassandra.yaml
4. Updated the microbenchmark to benchmark the read and write path for faster streaming as well as existing streaming
5. Addressed the dtest failures and added a dtest for faster streaming (https://github.com/apache/cassandra-dtest/compare/master...dineshjoshi:faster-streaming-rev2)

aweisberg commented:
The dtest doesn't validate that the kind of streaming you wanted to occur occured just that streaming succeeded.

I would also like to see it validate that if you run leveled compaction it will tend stream whole tables from the deeper levels which is one of the success criteria.

It would also be good to see that it works with CASSANDRA-6696 and size tiered and ends up streaming whole sstables.

dineshjoshi commented:
The dtest does validate the kind of streaming occurred see - https://github.com/apache/cassandra-dtest/compare/master...dineshjoshi:faster-streaming-rev2#diff-556858bf3074903213f73e382cf70297R58

aweisberg commented:
Ah it's the log statement block stream? Great. I think the thing I would look for then is just some log details about what percentage was block streamed vs regular streamed and assert that the expected amount of block streaming occurred.

aweisberg commented on src/java/org/apache/cassandra/db/DiskBoundaries.java:
```diff
@@ -129,4 +129,19 @@ public int getBoundariesFromSSTableDirectory(Descriptor descriptor)
+
+    private int getDiskIndex(DecoratedKey key)
+    {
+        int pos = Collections.binarySearch(positions, key);
```
OK so this is insertion point, and then -1 because the of 0 based indexing? Can you unit test this just to make sure that it find the correct spot?

I'm just not following the -1 I guess because Collections.binarySearch says insertion point -1 is what is returned. What if the key happens to be out of range of the ranges this node owns? This would go out of bounds?

dineshjoshi replied:
This method is pretty much a replica of `getDiskIndex(SSTableReader sstable)` (See: https://github.com/dineshjoshi/cassandra/blob/657ce3670a5291decf2939d72d51abe19333ef15/src/java/org/apache/cassandra/db/DiskBoundaries.java#L102) Instead of accepting a sstable, I pass in a `DecoratedKey`.

We ensure that the range is owned at the upper levels of the stack.

dineshjoshi commented on src/java/org/apache/cassandra/db/streaming/CassandraStreamHeader.java:
```diff
@@ -211,12 +228,16 @@ public void serialize(CassandraStreamHeader header, DataOutputPlus out, int vers
                 for (ComponentInfo info : header.components)
                     ComponentInfo.serializer.serialize(info, out, version);
+
+                ByteBufferUtil.writeWithShortLength(header.firstKey.getKey(), out);
```
Yes, this is the first situation where we're actually serializing `DecoratedKeys`. I am not sure if it'll overflow.

aweisberg commented on src/java/org/apache/cassandra/db/streaming/CassandraStreamHeader.java:
```diff
@@ -168,10 +232,35 @@ public void serialize(CassandraStreamHeader header, DataOutputPlus out, int vers
+                for (ComponentInfo info : header.components)
+                    ComponentInfo.serializer.serialize(info, out, version);
+
+                ByteBufferUtil.writeWithShortLength(header.firstKey.getKey(), out);
```
It's fine I forgot that the keys are hashes so they will always be fixed size and small.

aweisberg replied:
OK, actually Jeff pointed out that we do have partitioners where decorated keys are not hashes. So either we find out what the max length is or just use an int/varint and call it good. We don't send this often so it's fine.

aweisberg commented on src/java/org/apache/cassandra/net/async/RebufferingByteBufDataInputPlus.java:
```diff
@@ -249,4 +250,42 @@ public ByteBufAllocator getAllocator()
+                    throw new EOFException("EOF after " + copied + " bytes out of " + len);
+                }
+                if (buffer.remaining() == 0)
+                    return copied == 0 ? -1 : copied;
```
Heh, sorry I'm back for more. So reBuffering always does this check. I would change this to throw AssertionError because if it happens we don't actually know if it's end of stream or not because reBuffer didn't do what reBuffer is supposed to do.

dineshjoshi replied:
Right now the check is being made in `BigTableBlockWriter` and it throws an `IOException`, do we need to convert it into an `AssertionError`? We could convert this into an `EOFException` which is more uniform with the way the rest of the class handles things. WDYT?

aweisberg replied:
So reBuffer seems to have as its contract that it will throw EOF? That's why if you reach this point it is AssertionError since you don't know why reBuffer messed up.

Which check in BigTableBlockWriter are you talking about?

dineshjoshi replied:
Here - https://github.com/dineshjoshi/cassandra/blob/f9573ccef63535b787e91e46fa886f3e3ff2eee8/src/java/org/apache/cassandra/io/sstable/format/big/BigTableBlockWriter.java#L213

`reBuffer()` could likely fail due to network or channel being closed by the other side. It could be treated as a normal `IOException` as we're not guaranteed that we would get those many bytes from the other end. It feels a bit weird that rest of `RebufferingByteBufDataInputPlus` throws `EOFException` and this method would throw an `AssertionError` when there may be a network issue.

dineshjoshi replied:
Re-reading the code, the check on [line 227](https://github.com/dineshjoshi/cassandra/blob/f9573ccef63535b787e91e46fa886f3e3ff2eee8/src/java/org/apache/cassandra/net/async/RebufferingByteBufDataInputPlus.java#L277) is just being over cautious in case the `reBuffer` implementation changes and can return a empty buffer.

aweisberg replied:
If reBuffer() fails it's going to throw an IOException or RuntimeException that won't be caught there so it won't throw AssertionError. It will only throw AssertionError when reBuffer() fails to reBuffer() and also doesn't exit via an Exception. 

If reBuffer() doesn't reBuffer() and doesn't throw EOF then you don't know what's happening. That check doesn't do anything because if it doesn't read the require number of bytes it's going to throw EOF (in reBuffer()) and skip it anyways.

dineshjoshi replied:
Ok, makes sense. I have updated it to throw an `AssertionError`.

dineshjoshi committed:
Use variable length encoding for DecoratedKeys

aweisberg commented on src/java/org/apache/cassandra/db/streaming/CassandraOutgoingFile.java:
```diff
@@ -114,13 +153,54 @@ public void write(StreamSession session, DataOutputStreamPlus out, int version)
+            {
+                DecoratedKey key = iter.next();
+                boolean foundFlag = false;
+                for (Range<Token> r : requestedRanges)
```
So in the vnodes case is this going to be 256 ranges by default? Should this be a binary search instead of a scan?

dineshjoshi replied:
I refactored it to use `RangeOwnHelper`.

aweisberg commented on conf/cassandra.yaml:
```diff
@@ -784,6 +784,15 @@ compaction_throughput_mb_per_sec: 16
 # between the sstables, reducing page cache churn and keeping hot rows hot
 sstable_preemptive_open_interval_in_mb: 50
 
+# When enabled, permits Cassandra to zero copy eligible SSTables between
```
Worth documenting this optimization relies on having sstables that are wholly contained in the range being streamed and that configurations like size tiered without 6696 won't benefit, and would in fact spend time reading indexes checking to see if the optimization applies? Leveled should always benefit from this optimization.

dineshjoshi replied:
Leveled - yes, STCS - partly, 6696 should have a positive impact regardless of the strategy.

aweisberg replied:
I don't see how STCS can possibly benefit without 6696. It will always always have keys spanning all the ranges owned by that node in it.

aweisberg replied:
Except for the use cases where you have a small number of nodes, but that is a pretty big caveat.

dineshjoshi replied:
3 node cluster with RF=3?

aweisberg replied:
See above.

dineshjoshi replied:
Updated.

dineshjoshi committed:
Clean up code

dineshjoshi committed:
Throw AssertionError if reBuffer fails

dineshjoshi committed:
Update documentation

dineshjoshi committed:
Use RangeOwnHelper instead of custom logic

dineshjoshi commented:
@aweisberg I have updated the dtest to test with different compaction strategies and addressed remaining comments.

aweisberg commented:
Nice! Where is the dtest? The existing link doesn't seem to work and when I looked at your repo there are no branches with these tests?

iamaleksey commented on src/java/org/apache/cassandra/db/streaming/CassandraOutgoingFile.java:
```diff
@@ -114,13 +155,51 @@ public void write(StreamSession session, DataOutputStreamPlus out, int version)
+                try
+                {
+                    rangeOwnHelper.check(key);
+                } catch(RuntimeException e)
```
Catching `RuntimeException` is really not the way we should be using `RangeOwnHelper` here. Can you refactor `RangeOwnHelper` to introduce a method that would return a `boolean` instead?

iamaleksey replied:
On a more general note, this is potentially quite an expensive thing to do, especially for big sstables with skinny partitions, and in some cases this will introduce a performance regression.

The whole optimisation is realistically only useful for bootrstrap, decom, and rebuild, with LCS (which is still plenty useful and impactful and worth having). But it wouldn't normally kick in for regular repairs because of the full-cover requirement, and it won't normally kick in for STCS until CASSANDRA-10540 (range aware compaction) is implemented. In those cases having to read through the whole primary index is a perf regression that we shouldn't allow to happen.

The easiest way to avoid it would be to store sstable's effective token ranges in sstable metadata in relation to the node's ranges, making this check essentially free. Otherwise we should probably disable complete sstable streaming for STCS tables, at least until CASSANDRA-10540 is implemented. That however wouldn't address the regression to regular streaming, so keeping ranges in the metadata would be my preferred way to go.

aweisberg replied:
I mistakenly thought this index was a sampled index not a full index. Requiring a comparison of every partition key in every sstable for the entire data set seems like a big regression for some use cases. 

I was trying and failing to find the reasoning for why we switched to this.

dineshjoshi replied:
@iamaleksey thank you for the useful feedback. I did discuss this with @krummas and I believe while there was a room for improvement, the thinking back then was that the benefits would outweigh the cost. I looked through the codebase and this was the best way to definitely verify range containment as I was going for correctness. That said, what you suggest is obviously better. I am concerned about scope creep in this PR. Would it be ok if we address it as part of a separate PR?

It would also be useful, if we could design the effective range computation and storage in the metadata. I am not sure what sort of gotchas I might run into.

iamaleksey replied:
@dineshjoshi Duplicating a reply here: I think it's fine to not block this PR on the improvements, but disable zero copy for STCS for now, and commit that way. Once we optimise the containment check - ideally before 4.0 - enable for STCS as well.

Encoding information about the effective contained ranges is one way. Alternative way would be to invert the ranges we are checking for, and look for presence of keys in the gaps - but relying on summary and index to efficiently search rather than linearly scan. But again - let's temporarily disable zero copy for STCS for now, to not delay the ticket beyond necessary.

dineshjoshi replied:
I will disable Zero Copy streaming for STCS and I've created [CASSANDRA-14586](https://issues.apache.org/jira/browse/CASSANDRA-14586) to further discuss fixes.

iamaleksey commented on src/java/org/apache/cassandra/db/streaming/CassandraBlockStreamReader.java:
```diff
@@ -0,0 +1,184 @@
+    private static final Logger logger = LoggerFactory.getLogger(CassandraBlockStreamReader.class);
+    protected final TableId tableId;
+    protected final StreamSession session;
+    protected final int sstableLevel;
```
It has taken me some time (and @krummas's help) to prove that this wasn't a correctness issue, but at its best this is confusing/misleading code.

We extract `sstableLevel` from the header, but don't use it anywhere. Instead, since we stream `StatsMetadata` directly, we also inherit the level from there - regardless of whether `CassandraOutgoingStream.keepSSTableLevel` is set to `true`. If `LeveledManifest.canAddSSTable` check didn't exist, we'd be in trouble here. For clarity, I would probably look at that flag, and explicitly reset the level to `L0` if `keepSSTableLevel` is set to `false`.

P.S. What's the deal with all these `protected` fields?

dineshjoshi replied:
Cleaned up the protected fields. I still need to reset the level.

iamaleksey commented on src/java/org/apache/cassandra/config/DatabaseDescriptor.java:
```diff
@@ -2260,6 +2260,20 @@ public static int getStreamingConnectionsPerHost()
+    {
+        if (conf.server_encryption_options.enabled || conf.server_encryption_options.optional)
+        {
+            logger.debug("Internode encryption enabled. Disabling zero copy SSTable transfers for streaming.");
```
Nobody will ever see this at `debug` level. We should at minimum `warn` if `streaming_zerocopy_sstables_enabled` is and internode encryption are both enabled at the same time.

dineshjoshi replied:
Done

iamaleksey commented on src/java/org/apache/cassandra/db/streaming/ComponentInfo.java:
```diff
@@ -0,0 +1,97 @@
+               '}';
+    }
+
+    public boolean equals(Object o)
```
It's generally considered to be a bad practice to implement `equals()` and `hashCode()` unless that class is stored in a set or a map - or an upstream implementation of such. Otherwise it's just confusing boilerplate (confusing because it implies that the class is used in ways it clearly isn't).

In this case, there is a `List<ComponentInfo>` field in `CassandraStreamHeader`, which has an `equals()`/`hashCode()` implementation, which on the surface justifies these. But those, in turn, are actually dead code. So what we should do is remove the implementations of `equals()` and `hashCode()` here, and do the same in `CassandraStreamHeader`, being good citizens.

iamaleksey replied:
I shouldn't have said "it's generally considered" here, as apparently it's not quite, as @aweisberg pointed out. It's a subjective preference, and what I'd do, but feel free to disregard this particular comment.

dineshjoshi replied:
`CassandraStreamHeader`'s `equals()` and `hashCode()` method are not dead code. They're used in `CassandraStreamHeaderTest::serializerTest` which eventually ends up in [`SerializationUtils::assertSerializationCycle`](https://github.com/apache/cassandra/blob/9714a7c817b64a3358f69e536535c756c5c6df48/test/unit/org/apache/cassandra/serializers/SerializationUtils.java#L60). This is what originally forced me to implement those methods. The only other choice was to test the equality for `ComponentInfo` in `CassandraStreamHeader::equal()` which would have been a bad choice.

iamaleksey commented on src/java/org/apache/cassandra/db/streaming/ComponentInfo.java:
```diff
@@ -0,0 +1,97 @@
+import org.apache.cassandra.io.util.DataInputPlus;
+import org.apache.cassandra.io.util.DataOutputPlus;
+
+public class ComponentInfo
```
Now, if you look carefully at all current uses of `ComponentInfo`, you'll see that it's only ever being used as an element of lists, and never as a separate entity. As such it would be cleaner - and nicer to work with - to implement a class, say, `ComponentManifest` that would have an ordered map of `Component.Type` to `long` size, and expose the ordered keyset, with serializers for the whole manifest. As it's written now, ser/deser code is leaking outside, while it could instead be nicely encapsulated here.

dineshjoshi replied:
I have refactored this to `ComponentManifest`

iamaleksey commented on src/java/org/apache/cassandra/db/lifecycle/LogFile.java:
```diff
@@ -66,7 +66,7 @@
 
     // The transaction records, this set must be ORDER PRESERVING
-    private final LinkedHashSet<LogRecord> records = new LinkedHashSet<>();
+    private final Set<LogRecord> records = Collections.synchronizedSet(new LinkedHashSet<>()); // TODO: Hack until we fix CASSANDRA-14554
```
I'm not very well familiar with `LifeCycleTransaction` code, but I'm not sure this is sufficient as a workaround. @belliottsmith Could you please confirm that this - or isn't - sufficient? Here and the equivalent change to `LogReplicaSet`.

dineshjoshi replied:
@iamaleksey I don't think it is sufficient but we have to address this conclusively as part of 14554. I've already spoken to @belliottsmith regarding it and have a plan to fix this.

iamaleksey replied:
@dineshjoshi Ah. I see, that makes sense. The comment made it seem like this was supposed to be the fix for the issue, but this is just for the tests to pass? I'm good with that, given the extra context.

dineshjoshi replied:
I should've worded the comment differently but it is a hack to avoid a `ConcurrentModificationException` ([CASSANDRA-14554](https://issues.apache.org/jira/browse/CASSANDRA-14554)). This currently exists in trunk and was not introduced as part of this PR hence the separate ticket.

dineshjoshi committed:
change log level to warn

dineshjoshi committed:
Refactor ComponentInfo -> ComponentManifest

dineshjoshi committed:
Augment RangeOwnHelper to return a boolean

dineshjoshi committed:
Make fields final

dineshjoshi commented:
@iamaleksey I have resolved most of your comments. I still have a couple to go. I will update this PR when I am done with those and get a clean dtest run.

dineshjoshi committed:
Limit faster streaming to LCS. Reset level information using CassandraStreamHeader

dineshjoshi commented on src/java/org/apache/cassandra/db/streaming/CassandraOutgoingFile.java:
```diff
@@ -62,7 +90,23 @@ public CassandraOutgoingFile(StreamOperation operation, Ref<SSTableReader> ref,
+    public static ComponentManifest getComponentManifest(SSTableReader sstable)
+    {
+        LinkedHashMap<Component.Type, Long> components = new LinkedHashMap<>(STREAM_COMPONENTS.size());
+        for (Component component : STREAM_COMPONENTS)
```
@iamaleksey would it be worth having a list of required components here? Any one of the missing components will be a fatal? For example, when I reset the level on the receiving side, I expect stats to exist and is sent over.

dineshjoshi commented:
@iamaleksey I *think* I've resolved all your comments. The last [dtest](https://circleci.com/gh/dineshjoshi/workflows/cassandra/tree/14556-trunk-optimize-streaming) was all green. I'm not sure about the latest push but I don't forsee any breakages. Please let me know if there are more changes required.

iamaleksey committed:
Rename checkBoolean

iamaleksey committed:
Propagate result of writer.abort(), replaced deprecated
Throwables.propagate() call

iamaleksey committed:
CBSR: remove redundant fields, make logging consistent (esp. re: prettyPrintMemory() use)

iamaleksey commented on src/java/org/apache/cassandra/db/streaming/CassandraBlockStreamReader.java:
```diff
@@ -0,0 +1,177 @@
+                        session.planId(), cfs.keyspace.getName(), cfs.getTableName(), e);
+            if (writer != null)
+            {
+                writer.abort(e);
```
We shouldn't be ignoring the result of `writer.abort()`, and propagate it instead of e.

dineshjoshi replied:
Cherry picked this change.

iamaleksey commented on src/java/org/apache/cassandra/db/streaming/CassandraBlockStreamReader.java:
```diff
@@ -0,0 +1,177 @@
+            {
+                writer.abort(e);
+            }
+            throw Throwables.propagate(e);
```
`Throwables.propagate()` is deprecated now, so we should probably not be using it in new code. Guava developers on why - [here](https://github.com/google/guava/wiki/Why-we-deprecated-Throwables.propagate).

dineshjoshi replied:
Cherry picked. Appreciate the link.

iamaleksey commented on src/java/org/apache/cassandra/io/sstable/format/big/BigTableBlockWriter.java:
```diff
@@ -0,0 +1,221 @@
+    public Throwable commit(Throwable accumulate)
+    {
+        for (SequentialWriter writer : componentWriters.values())
+            writer.commit(accumulate);
```
We shouldn't be ignoring the result of writer.commit(), and should return it instead of accumulate.

dineshjoshi replied:
Cherry picked.

iamaleksey commented on src/java/org/apache/cassandra/io/sstable/format/big/BigTableBlockWriter.java:
```diff
@@ -0,0 +1,221 @@
+        for (SequentialWriter writer : componentWriters.values())
+            writer.abort(accumulate);
+
+        return accumulate;
```
We shouldn't be ignoring the result of writer.abort(), and should return it instead of accumulate.

dineshjoshi replied:
Cherry picked.

iamaleksey commented on src/java/org/apache/cassandra/db/compaction/Verifier.java:
```diff
@@ -361,12 +361,26 @@ public RangeOwnHelper(List<Range<Token>> normalizedRanges)
          */
         public void check(DecoratedKey key)
+        {
+            if (!checkBoolean(key))
```
Yep, this is exactly what I had in mind, thanks. Not sure of the names, though. Would prefer the throwing `check()` to be renamed into `validate()`, and `checkBoolean()` to become `check()`. `validate()` is the name we often give to such methods (that throw on failure).

dineshjoshi replied:
I am cherry picking this change. 

The only concern I have is we're now changing the signature and behavior of a pre-existing method. I generally try to avoid such changes to avoid confusing other authors.

iamaleksey replied:
Please don't avoid it - unless it's in public API. Change as many layers as necessary, in general, until the new code fits into the big picture. We allow and welcome it here.

dineshjoshi replied:
Thanks, I will bear that in mind for future patches.

iamaleksey commented on src/java/org/apache/cassandra/db/streaming/CassandraBlockStreamReader.java:
```diff
@@ -0,0 +1,177 @@
+            // we should only ever be streaming pending repair
+            // sstables if the session has a pending repair id
+            if (!session.getPendingRepair().equals(header.pendingRepair))
+                throw new IllegalStateException(String.format("Stream Session & SSTable ({}) pendingRepair UUID mismatch.",
```
{} isn't a valid placeholder for `String.format()`. Maybe this used to be a logging call?

dineshjoshi replied:
Great catch. Yes, it used to be a logging call.

iamaleksey commented on src/java/org/apache/cassandra/db/streaming/CassandraBlockStreamReader.java:
```diff
@@ -0,0 +1,177 @@
+    private final TableId tableId;
+    private final StreamSession session;
+    private final int sstableLevel;
+    private final SerializationHeader.Component header;
```
This is an unused field. We'd be better off storing the `CassandraStreamHeader` here, though, and getting rid of all the redundant extracted fields (`version`, `firstKey`, etc).

dineshjoshi replied:
Cherry picked.

iamaleksey commented on src/java/org/apache/cassandra/db/streaming/CassandraBlockStreamReader.java:
```diff
@@ -0,0 +1,177 @@
+        if (cfs == null)
+        {
+            // schema was dropped during streaming
+            throw new IOException("CF " + tableId + " was dropped during streaming");
```
We generally don't call things `CF` in any new code anymore, `table` is the preferred terminology.

dineshjoshi replied:
Cherry picked

iamaleksey commented on src/java/org/apache/cassandra/db/streaming/CassandraBlockStreamReader.java:
```diff
@@ -0,0 +1,177 @@
+        }
+
+        logger.debug("[Stream #{}] Start receiving file #{} from {}, size = {}, ks = '{}', table = '{}'.",
+                     session.planId(), fileSeqNum, session.peer, totalSize, cfs.keyspace.getName(), cfs.getTableName());
```
Should be using `prettyPrintMemory()` here for `totalSize` too. Also would be cleaner if we static-import it, as it's being used a lot? Also, `TableMetadata.toString()` is overridden nicely to return keyspace.table string, just for cases like this.

dineshjoshi replied:
Cherry picked.

iamaleksey commented on src/java/org/apache/cassandra/io/sstable/format/big/BigTableBlockWriter.java:
```diff
@@ -0,0 +1,221 @@
+              DatabaseDescriptor.getDiskOptimizationStrategy());
+        txn.trackNew(this);
+        this.metadata = metadata;
+        this.txn = txn;
```
The field is unused.

dineshjoshi replied:
Removed.

iamaleksey commented on src/java/org/apache/cassandra/io/sstable/format/big/BigTableBlockWriter.java:
```diff
@@ -0,0 +1,221 @@
+        this.componentWriters = new HashMap<>(components.size());
+
+        assert supportedComponents.containsAll(components) : String.format("Unsupported streaming component detected %s",
+                                                                           new HashSet(components).removeAll(supportedComponents));
```
`Set.removeAll()` returns a `boolean`, so this won't quite work unfortunately.

dineshjoshi replied:
Fixed.

iamaleksey commented on src/java/org/apache/cassandra/io/sstable/format/big/BigTableBlockWriter.java:
```diff
@@ -0,0 +1,221 @@
+        txn.trackNew(this);
+        this.metadata = metadata;
+        this.txn = txn;
+        this.componentWriters = new HashMap<>(components.size());
```
Can be a `EnumMap` (:

iamaleksey commented on src/java/org/apache/cassandra/io/sstable/format/big/BigTableBlockWriter.java:
```diff
@@ -0,0 +1,221 @@
+    private volatile SSTableReader finalReader;
+    private final Map<Component.Type, SequentialWriter> componentWriters;
+
+    private final Logger logger = LoggerFactory.getLogger(BigTableBlockWriter.class);
```
Logger field should be static.

iamaleksey commented on src/java/org/apache/cassandra/io/sstable/format/big/BigTableBlockWriter.java:
```diff
@@ -0,0 +1,221 @@
+
+    private final Logger logger = LoggerFactory.getLogger(BigTableBlockWriter.class);
+
+    private final SequentialWriterOption writerOption = SequentialWriterOption.newBuilder()
```
Presumably should be static too.

iamaleksey commented on src/java/org/apache/cassandra/io/sstable/format/big/BigTableBlockWriter.java:
```diff
@@ -0,0 +1,221 @@
+
+    public void writeComponent(Component.Type type, DataInputPlus in, long size)
+    {
+        logger.info("Writing component {} to {} length {}", type, componentWriters.get(type).getPath(), size);
```
Presumably should be using `prettyPrintMemory()`?

iamaleksey commented on src/java/org/apache/cassandra/io/sstable/format/big/BigTableBlockWriter.java:
```diff
@@ -0,0 +1,221 @@
+            long bytesWritten = in.consumeUntil(writer, size);
+
+            if (bytesWritten != size)
+                throw new IOException(String.format("Failed to read correct number of bytes from Channel {}", writer));
```
{} is not a valid placeholder for `String.format()`.

dineshjoshi replied:
For all of `BTBW` I picked the change from your branch with just a couple minor changes.

iamaleksey reviewed (changes requested):
Thanks. It's surely looking better now, though we still have some work to do.

One of the things that haven't been addressed is the counter sstable streaming issue (see my initial JIRA comment). TL;DR we shouldn't enable full-sstable streaming if it's a counter sstable with legacy shards.

To speed things up, here is my review branch: 
https://github.com/iamaleksey/cassandra/commits/14556-trunk-optimize-streaming-review. Please feel free to cherry-pick the commits from here as yours, or just copy them, or merely use as illustration of what I have in mind.

iamaleksey commented on src/java/org/apache/cassandra/db/streaming/ComponentManifest.java:
```diff
@@ -0,0 +1,129 @@
+
+public class ComponentManifest
+{
+    private final LinkedHashMap<Component.Type, Long> manifest;
```
Instead of having a map of type to size with a separate set of components, it'd be nicer to just have a map of `Component` -> `Long`, I think. One is definitely unnecessary. Nor do we save a lot be caching the total size.

dineshjoshi replied:
I had originally thought of going in this direction. I have rewritten this class and incorporated your suggestion including most stylistic changes.

iamaleksey commented on src/java/org/apache/cassandra/db/streaming/ComponentManifest.java:
```diff
@@ -0,0 +1,129 @@
+public class ComponentManifest
+{
+    private final LinkedHashMap<Component.Type, Long> manifest;
+    private final Set<Component> components = new LinkedHashSet<>(Component.Type.values().length);
```
`.values()` allocates an array on each call, so any benefit we get from presizing the hashset is probably lost. Nor does it really matter here (:

dineshjoshi replied:
Done.

iamaleksey commented on src/java/org/apache/cassandra/db/streaming/ComponentManifest.java:
```diff
@@ -0,0 +1,129 @@
+
+        private void serialize(Component.Type type, long size, DataOutputPlus out) throws IOException
+        {
+            out.writeByte(type.id);
```
Encoding types as a single byte reduces the size of the message a little, but it has a cost of unnecessarily increasing complexity. We do this in `Verb`, but there it is for a good reason: we ser/deser `Verb`s *a lot*. Here we'd be better of with just relying on `valueOf()`, IMO.

dineshjoshi replied:
@jasobrown and I debated this. Basically, we chose to go with the byte approach as it seemed to be better than going with ordinals. In degenerate cases where we have a large number of SSTables, this will make a difference as we will be sending a lot of these messages around. unless you strongly oppose removing this, I am inclined to leaving this as-is.

iamaleksey reviewed (commented):
See some comments regarding `ComponentManifest` below. I mocked up changes to it here - 
https://github.com/iamaleksey/cassandra/commit/200d43652d95276f45d44e45d6ece8c0715ff636, for illustrative purposes.

The commit has some purely stylistic changes in addition to addressing the issues raised below - it's what I would've done, and I believe it makes the code a bit cleaner and nicer, but please feel free to ignore those entirely if you disagree strongly.

iamaleksey commented on src/java/org/apache/cassandra/db/streaming/CassandraBlockStreamWriter.java:
```diff
@@ -0,0 +1,109 @@
+                         sstable.descriptor.generation, component, session.peer, FBUtilities.prettyPrintMemory(bytesRead),
+                         FBUtilities.prettyPrintMemory(length), FBUtilities.prettyPrintMemory(totalSize));
+
+            byteBufDataOutputStreamPlus.flush();
```
Do we really need to flush the stream manually? Either way, we should definitely not do this inside this loop. Move it outside or remove completely?

dineshjoshi replied:
I just have it as a good measure to ensure that when `write()` returns, the stream is flushed. I have moved it outside the loop.

iamaleksey commented on src/java/org/apache/cassandra/db/streaming/CassandraBlockStreamWriter.java:
```diff
@@ -0,0 +1,109 @@
+
+            byteBufDataOutputStreamPlus.flush();
+
+            logger.debug("[Stream #{}] Finished block streaming sstable {} to {}, xfered = {}, totalSize = {}",
```
I believe this was supposed to go outside of the for-loop as well (:

dineshjoshi replied:
Done.

iamaleksey commented on src/java/org/apache/cassandra/db/streaming/CassandraBlockStreamWriter.java:
```diff
@@ -0,0 +1,109 @@
+    {
+        long totalSize = manifest.getTotalSize();
+        logger.debug("[Stream #{}] Start streaming sstable {} to {}, repairedAt = {}, totalSize = {}", session.planId(),
+                     sstable.getFilename(), session.peer, sstable.getSSTableMetadata().repairedAt, totalSize);
```
Missed `prettyPrintMemory()` on `totalSize`?

dineshjoshi replied:
Fixed.

iamaleksey commented on src/java/org/apache/cassandra/db/streaming/CassandraBlockStreamWriter.java:
```diff
@@ -0,0 +1,109 @@
+            long length = in.size();
+
+            // tracks write progress
+            long bytesRead = 0;
```
Is this a variable left from some previous iteration of the class? Because we only update it once.. perhaps should be called `bytesWritten` (I assume it was copy-pasted from the read impl?)

dineshjoshi replied:
I did rename this.

iamaleksey commented on src/java/org/apache/cassandra/db/streaming/CassandraBlockStreamWriter.java:
```diff
@@ -0,0 +1,109 @@
+            logger.debug("[Stream #{}] Block streaming {}.{} gen {} component {} size {}", session.planId(),
+                        sstable.getKeyspaceName(), sstable.getColumnFamilyName(), sstable.descriptor.generation, component, length);
+
+            bytesRead += byteBufDataOutputStreamPlus.writeToChannel(in, limiter);
```
So this should, essentially, be same as `length`, correct?

dineshjoshi replied:
Yes, this *should* be the same. I avoided making that assumption and instead use what `writeToChannel()` returns. It is also used to update the `session.progress()`. Ideally, `bytesRead == length`, else we have some bug. I did rewrite this to make it concise.

dineshjoshi committed:
Use String format specifier

dineshjoshi committed:
BTBW use static imports, fix assertion logic and reformat

dineshjoshi committed:
Address code review comments. Refactor ComponentManifest.

dineshjoshi committed:
COF - Do not use zero copy path if legacy counter shards are present

dineshjoshi committed:
Update documentation

dineshjoshi committed:
Adjust unit test

dineshjoshi committed:
Rename variable to more accurate reflect its purpose

dineshjoshi committed:
COF - Refactor - lazy evaluate expensive containment and some stylistic changes

dineshjoshi commented:
@iamaleksey I've addressed your comments including the one about disabling faster streaming for legacy counter shards.

I did add a much less expensive check for STCS. It won't get all SSTables accurately but it is way cheaper than what I have for LCS. Let me know your thoughts.

iamaleksey commented on src/java/org/apache/cassandra/io/sstable/format/big/BigTableBlockWriter.java:
```diff
@@ -48,51 +47,61 @@
-                                                                           new HashSet(components).removeAll(supportedComponents));
+        if (!SUPPORTED_COMPONENTS.containsAll(components))
+            throw new AssertionError(format("Unsupported streaming component detected %s",
+                                            Sets.difference(components, SUPPORTED_COMPONENTS)));
```
Neat. I either forgot, or didn't know that `Sets.difference()` was a thing. This is nicer than the way I proposed (:

iamaleksey commented on src/java/org/apache/cassandra/db/streaming/CassandraStreamHeader.java:
```diff
@@ -183,9 +261,26 @@ public CassandraStreamHeader deserialize(DataInputPlus in, int version) throws I
+                ByteBuffer keyBuf = ByteBufferUtil.readWithVIntLength(in);
+                IPartitioner partitioner = partitionerMapper.apply(tableId);
+                if (partitioner == null)
+                    throw new IllegalArgumentException(String.format("Could not determine partitioner for tableId {}", tableId));
```
Another instance of `String.format()` format string with `{}` instead of `%s`, looks like.

dineshjoshi replied:
Fixed.

iamaleksey commented on test/unit/org/apache/cassandra/db/streaming/CassandraStreamHeaderTest.java:
```diff
@@ -43,8 +51,38 @@ public void serializerTest()
+        String ddl = "CREATE TABLE tbl (k INT PRIMARY KEY, v INT)";
+        TableMetadata metadata = CreateTableStatement.parse(ddl, "ks").build();
+
+        ComponentManifest manifest = new ComponentManifest(new HashMap(ImmutableMap.of(Component.DATA, 100L)));
```
No need to wrap the immutable map in a hashmap here (:

dineshjoshi replied:
Fixed. Not sure why I did this in the first place.

iamaleksey committed:
Introduce a Builder to CassandraStreamHeader

also:
- better naming
- fix a bug in CassandraStreamHeader.serializedSize()
- remove objects without a toString() implementation from CSH.toString()

iamaleksey commented on src/java/org/apache/cassandra/db/streaming/CassandraBlockStreamWriter.java:
```diff
@@ -0,0 +1,122 @@
+                         sstable.getKeyspaceName(),
+                         sstable.getColumnFamilyName(),
+                         sstable.descriptor.generation,
+                         component, length);
```
`prettyPrintMemory()` missing here for `length`.

dineshjoshi replied:
Fixed.

iamaleksey reviewed (commented):
Some minor things below (one of them a bug).

iamaleksey commented on src/java/org/apache/cassandra/db/streaming/CassandraStreamHeader.java:
```diff
@@ -65,18 +85,43 @@ private CassandraStreamHeader(Version version, SSTableFormat.Type format, long e
     }
 
-    public CassandraStreamHeader(Version version, SSTableFormat.Type format, long estimatedKeys, List<SSTableReader.PartitionPositionBounds> sections, CompressionMetadata compressionMetadata, int sstableLevel, SerializationHeader.Component header)
+    private CassandraStreamHeader(Version version, SSTableFormat.Type format, long estimatedKeys,
```
The introduction of the new fields and constructors got us to 5 constructors total with up to 10 arguments, which is no longer manageable, and calls for a builder. It's boring and tedious work, so I did it myself and pushed here - 
https://github.com/iamaleksey/cassandra/commit/321d21747faa46afcf34518ebdeb811f2a805de8 - please feel free to cherry-pick.

In addition to introducing the builder, the commit renames `fullStream` to something a bit more meaningful (`isEntireSSTable`) that clearly reflects what's actually happening, fixes a bug in `serializedSize()` where compression info isn't initialized, and removes some fields without `toString()` implementations from header's own `toString()`.

dineshjoshi replied:
I have cherry picked this change. Thanks!

dineshjoshi committed:
add pretty printing to missing field. fix format specifier. formatting

dineshjoshi committed:
Make naming consistent

iamaleksey commented on src/java/org/apache/cassandra/db/streaming/ComponentManifest.java:
```diff
@@ -0,0 +1,130 @@
+            out.writeUnsignedVInt(manifest.components.size());
+            for (Map.Entry<Component, Long> entry : manifest.components.entrySet())
+            {
+                out.writeByte(entry.getKey().type.id);
```
Talked to @dineshjoshi offline, and we realised that this is incomplete - and neither was my proposed version. For completeness, when want to serialize the whole component info, not just its type. And it has two important fields - type and name. Name will usually be derived from the type, but not always. And even though we don't support streaming those components (custom and SI), we might want to change it in the future, and the protocol should allow it.

So I suggest we encode`component.type.name()`, the full enum name, followed by `component.name()`. It's a little heavier, but this is completely irrelevant in the big picture, size-wise.

The upside is that we can handle encode/decode any component necessary in the future, loss-free. And, again, we don't really need to assign ids. `valueOf()` is plenty good, and allows extension without overlap risk like in `Verb`.

iamaleksey replied:
FWIW, I realize that for most components this will be a bit redundant. Technically it's sufficient to just store `component.name`, and get the full `Component` via `Component.parse()`. If you don't like redundancy and want to do it that way, that's perfectly fine too - I'm cool with either option.

dineshjoshi replied:
Done. I'm just using `component.name`. I think this should be sufficient for this PR's scope.

iamaleksey commented on src/java/org/apache/cassandra/db/streaming/CassandraBlockStreamReader.java:
```diff
@@ -0,0 +1,173 @@
+                     prettyPrintMemory(totalSize),
+                     cfs.metadata());
+
+        BigTableBlockWriter writer = null;
```
It's likely that we'll have more formats in near future. I don't suggest we make this generic, until then, but for now at least validate that the format in the header is `BIG` perhaps?

dineshjoshi replied:
Done

dineshjoshi committed:
Make sure we disable zero copy and warn once.

dineshjoshi committed:
Stylistic changes

dineshjoshi committed:
Remove IStreamWriter

dineshjoshi committed:
Assert format

dineshjoshi committed:
Stylistic changes

dineshjoshi committed:
Simplify ComponentManifest ser/des

dineshjoshi commented:
@iamaleksey made a few more changes - 

1. Got rid of `IStreamWriter`
2. Ensured we're logging the configuration warning only once at start up iff zero copy streaming is enabled
3. Few stylistic changes

iamaleksey commented:
@dineshjoshi nice round of changes! I think we are there, or almost-almost there, at worst. Let me do one last pass before we go ahead and commit.

iamaleksey commented on src/java/org/apache/cassandra/db/streaming/CassandraBlockStreamWriter.java:
```diff
@@ -0,0 +1,121 @@
+
+import org.apache.cassandra.io.sstable.Component;
+import org.apache.cassandra.io.sstable.format.SSTableReader;
+import org.apache.cassandra.io.util.DataOutputStreamPlus;
```
Nit: import now unused.

iamaleksey commented on src/java/org/apache/cassandra/db/streaming/CassandraOutgoingFile.java:
```diff
@@ -18,51 +18,102 @@
  */
 public class CassandraOutgoingFile implements OutgoingStream
 {
+    private static final boolean isZeroCopySSTableTransfersEnabled = DatabaseDescriptor.isZeroCopySSTableTransfersEnabled();
```
Can get rid of this field now that `DatabaseDescriptor.isZeroCopySSTableTransferEnabled()` just reads a field from `Config`.

iamaleksey commented on src/java/org/apache/cassandra/db/streaming/CassandraOutgoingFile.java:
```diff
@@ -114,11 +165,77 @@ public void write(StreamSession session, DataOutputStreamPlus out, int version)
+            return (ranges != null
+                        && ranges.size() == 1
+                        && ranges.get(0)
+                                 .contains(new Range<>(ref.get().first.getToken(),
```
Should be Bounds for the sstable, since range is (first, last] and sstable contains [first, last] (thanks @krummas). But on a higher level note, maybe just stick to LCS only for now, and enable STCS after we commit CASSANDRA-14568? We are going to do that in 4.0, so this temporary special case won't make it to a release anyway ¯\_(ツ)_/¯

iamaleksey reviewed (approved):
Tiny last round - 2 nits, 1 minor bug/suggestion.

+1 overall - although might want to make some minor renames as discussed offline before or on commit.

dineshjoshi committed:
Remove unused import, STCS check. Call into DatabaseDescriptor directly instead of caching value.

dineshjoshi committed:
Better naming

dineshjoshi committed:
Slight renaming for consistency

dineshjoshi committed:
Update logging

dineshjoshi committed:
Make naming in cassandra.yaml and database descriptor consistent
